language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -588,6 +309,11 @@ impl AbstractDomain for AbstractState {
         {
             return JoinResult::Error;
         }
+        if Self::borrowed_value_unavailable(self, state)
+            || Self::borrowed_value_unavailable(state, self)
+        {
```

Comment:
This feels way way too restrictive to me. We should never give a JoinResult::Error. 

The new borrow graph formulation can definitely handle a more expressive join. Perhaps this is more apparent if we split the type_and_memory_safety pass into 3 passes
- Type safety
- Locals safety
- Borrow safety

Reply:
JoinResult::Error is not needed only for borrow checking.  Consider a program without references.  Suppose x is a local variable holding a resource.  If the analysis arrives at a join point with x available on one side and unavailable on the other, I do not see any option other than giving an error.  What do you suggest should happen in this case?

Reply:
We have discussed this before. 

You assign it a third state of "Maybe assigned" a value. And then error in the same cases as "Assigned" for a resource, but treat it as "unassigned" for an unrestricted value

https://github.com/libra/libra/blob/master/language/move-lang/src/cfgir/locals/state.rs#L12#L19

Reply:
I think this is more a matter of taste.  I don't find eliminating JoinError in favor of a Maybe either more principled or easier to understand.

Reply:
It's not just about taste. It forces you to make a more expressive abstract state, and to think about cases more carefully, as you cannot fall onto join error as a crutch. 

I think its a style choice that leads to designs that are not fully expressive.

Reply:
The join error is not a crutch.  My understanding is that we will report an error in every single case when a join error happens.

Reply:
Discussed offline also.  JoinError appears to be fine.